
# 前端错误基础
![](https://tva1.sinaimg.cn/large/007S8ZIlly1ghlz7kwkccj30u014f0zr.jpg)

---

## 前端错误的类型

在日常开发中，我们可能会碰到什么类型的错误呢？归纳起来主要有以下几种：

### 1. ECMAScript exceptions

### 2. DOMException and DOMError

### 3. 网络静态资源加载错误

### 4. 跨域引用script导致的script error

### 5. 页面崩溃

---

### 1. ECMAScript exceptions

- **ReferenceError**

这类型错误通常是访问那些**未经声明或者不存在的变量**时所触发的错误。比如：

```js
console.log(a);
```

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghknf2g2c4j312c02mwf7.jpg)

- **SyntaxError**

当我们JavaScript代码中出现语法错误的时候，就会导致这种类型的错误。比如：

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghko9er56uj312i06agmp.jpg)

- **TypeError**

我们在运行时看到最多的错误类型了。因为JavaScript是动态弱类型，所有变量的数据类型都出于变动中。如果，运行时期间，**在不恰当的数据类型变量上执行不恰当的操作的时候**，浏览器就会报这个大名鼎鼎的TypeError。

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghkosqfycfj312m098myx.jpg)

- **RangeError**

当我们在初始化数组的时候，假如我们申请的长度过长的话，就会引发这种错误。也就是说，当数值超出相应范围的时候，就会触发这种错误。比如：

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghkovh1kxhj31280ie0wy.jpg)

- **URIError**

在使用encodeURI()、encodeURIComponent或者decodeURI()、decodeURIComponent, 当传入的字符换的URL的格式不对的时候，就会导致URIError。比如:

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghkp4589orj312803s751.jpg)

---

### 2. DOMException and DOMError

DOMException就是我们在调用web API方法或者属性的时候所发生的错误,或者简单地说是在执行DOM操作的时候所抛出的错误。[DOMError]([https://developer.mozilla.org/zh-CN/docs/Web/API/DOMError](https://developer.mozilla.org/zh-CN/docs/Web/API/DOMError))不是标准规范，这里就没必要讨论它了。

在跟音视频打交道的过程中，在不适当的时机调用web API方法的时候，往往会出现这类型的错误。比如：

```js
window.onload = function () {
    var video = document.getElementById('video');
    video.play();
};

<video id="video" preload="none" src="http://vfx.mtime.cn/Video/2019/02/04/mp4/190204084208765161.mp4"></video>
```

上面代码会报一个[DOMException](https://developers.google.cn/web/updates/2017/06/play-request-was-interrupted?hl=en#fix):

> Uncaught (in promise) DOMException: play() failed because the user didn't interact with the document first.

又比如在错误地调用了DOM接口:

```js
<!DOCTYPE html>
<html>
<head>
</head>
<body>
     <script type="text/javascript">
        function ThrowDOMException () {

            var elem = document.createAttribute ("123");

        }
    </script>

    <button onclick="ThrowDOMException ()">Throw a DOM exception</button>

</body>
</html>
```

上面的示例，浏览器会报一个[DOMException](http://help.dottoro.com/ljushpcx.php;):

> Uncaught DOMException: Failed to execute 'createAttribute' on 'Document': The localName provided ('123') contains an invalid character.

### 3. 网络静态资源加载错误

常见的网络静态资源包括 html文件，css文件，javascript文件，图片，音频，视频，iframe等等。所有的网络静态资源都有可能遇到加载错误这个问题。加载错误的原因可能是：

- url写错了
- 服务器根本没有这个资源
- 服务器出现内部错误
- 网络忙，请求超时而导致资源加载失败

不管什么原因，凡是客户端没有加载成功的，一律视为出现了加载错误。

### 4. script error

当跨域引用了另外一个域下JavaScript资源，并且这个JavaScript执出错的情况下，浏览器就会抛出这个错误。

假如，我们在`another-domain.com`域名下有这么一个js文件:

```js

const a = {};

console.log(a.b.c);

```

在我们自己的域名`origin-domain.com`中，我们去引用这个文件:

```js


<script src="http://another-domain.com/index.js"></script>

```

那么在这种场景下，浏览器就会抛出一个`script error`类型的错误。`script error`类型的错误基本上是在告诉你，你跨域引用的脚本执行出错了，但是你没有知道具体错误信息的权利。一切的一切都是因为，浏览器为了信息安全，制定了所谓的**同源策略**（`same-origin policy`）。

### 5. page crash

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghkyfx0zovj30wc0cq765.jpg)

---

## 前端错误的防范

DOMException一般是在你对DOM API理解或者记错的情况下犯的错误。在出错前，你潜意识地就觉得自己可能是对。所以在这种情况下，是无法人为地防范的。其他的几种错误，错误源头几乎不在你的掌控范围内，所以，你也无法防范。

接下来主要探讨一下我们该如何防范ECMAScript exceptions这种类型的错误，这是我们能人为防范的。

其实，若引入TypeScript，为我们的JavaScript加上一层类型的保护衣后，这里面的很多错误都能够在代码编写期得到提醒和纠正。但是，这一小节还是想来谈谈，如何通过培养一些好的JavaScript编程习惯来防范这类型的错误。

在ECMAScript exceptions这类型的错误中，我们主要防范的是以下三种子类型的错误：

- 隐性类型转换导致的错误
- 缺乏类型防守导致的错误
- 通信错误

### `隐性类型转换`导致的错误

这种错误一般在对隐性类型转换不够警惕而导致的错误。

举个例子，你手头上有一个数字类型的id，然后从后端接口请求回一个列表。如果列表中某个item的id等于你手头这个id的话，那么你就想把这个item的的边框高亮起来。这个时候你考虑到后端同事的不靠谱，有时候返回字符串类型，有时候返回数字类型，你于是这么写：

```js
this.list = this.list.map(item=> {
    if(item.id == myID){
        return {
            ...item,
            iSelected: true
        }
    }

    return item;
})
```

是的，这里为了容错性，使用了等于号所带来的`隐性类型转换`而不是全等号也算是用得其所。所以说，隐性类型转换没有错，关键是你知道怎么用。但是话说回来，隐形类型转换终究是会降低了代码的可预测性，使得代码难以维护。

**`众所周知，当使用相等(==)和不相等(!=)操作符，或者在if,for以及while等流程控制语句使用非布尔值类型的的数据的时候，实质上我们是在使用隐性类型转换`**。

当你对隐性类型转换的警惕性稍有下降的时候，你就会犯错了。

举个例子，在理财业务开发中，我们有个输入框，它主要是用于输入购买金额。在输入框的下面，我们会有一个提示语。实时地跟进用户的输入同步计算该购买金额所产生的利息：

```js
let hintText = '请输入购买金额';
const purchaseAmount = parseFloat(document.getElementById('input').value) ;

if(purchaseAmount) {
    const income = .... // 做算术四则运算;
    hintText = `预计收益是$(income)元`;
}
```

因为用户的有可能输入的是“0”，这个时候因为if语句的隐性类型转换为false，导致无法更新hintText。结果是，用户虽然是输入了东西，但是界面还是显示默认的提示语“请输入购买金额”。虽然不是一个致命的错误，但是很明显是一个不符合逻辑的错误。

此处，如果能够避免隐性类型转换，那么我们就不会犯下这个错误：

```js
if(typeof purchaseAmount === 'number' && !isNaN(purchaseAmount)){
    // ......
}
```

虽然此处的错误不会导致js代码的运行时错误，但是保不齐下次就会把不恰当的类型漏放进去到if语句的代码块中，然后基于在此类型数据做不恰当的操作，很有可能就导致一个TypeError的错误。

所以，我们要避免使用隐性转换。怎样避免法呢？

- 使用全等号（===）和不全等号(!==)来代替等号(==)和不等号(!=);
- 在各种流程控制语句中，做条件判断的时候要确保传入的是一个布尔值（或者值为布尔值的表达式）。

### 缺乏`类型防守`导致的错误

JavaScript是一个动态弱类型的编程语言。所以，在使用JS进行软件开发的过程中，为了保证程序的健壮性，类型防守太重要了。

举个例子,下面我们要实现一个从url上获取query string的函数：

```js
function getQueryString(url){
    const pos = url.indexOf('?');
    if (pos > -1) {
        return url.slice(pos + 1);
    }

    return '';
}
```

我们理所当然地以为url是字符串类型。但是万一调用我们函数的人，一不小心传了数字类型值进来呢？显然，此处，在不恰当的数据类型上调用了不恰当的方法，浏览器会在运行时给我们一个大大的“`Uncaught TypeError:url.indexOf is not a function`”。只要我们提高警惕性，加上一条简单的类型防守的判断语句即可避免犯下这个错误：

```js
function getQueryString(url){
    if(typeof url === 'string') {
        const pos = url.indexOf('?');
        if (pos > -1) {
            return url.slice(pos + 1);
        }
    }

    return '';
}
```

### 通信错误

在前端开发的过程总，一般接触到的通信类型包括：前端与客户端的通信，前端与服务端的通信。

在前端与客户端的通信中，一般出现得比较多的是前端传递给客户端一个未经编码的url，导致客户端解析出错。比如：

```js
hdf://login?redirect=https://frontend.com/index.html?a=b&c=d
```

上面这个问题，我们使用encodeURIComponent()方法对“redirect”后面的字符串进行编码一下就可以解决这个问题：

```js
hdf://login?redirect=https%3A%2F%2Ffrontend.com%2Findex.html%3Fa%3Db%26c%3Dd
```

而客户端那边在解析之前也需要用匹配的方法来进行解码。

而在跟服务端通信的过程中，不少值得注意的事情。很多时候，前后端已经通过接口文档规定好数据类型和结构了，但是后端还是犯错，没有按照约定来返回数据类型和结构。

关于数据类型,后端常常犯的错有：

- 数字类型与字符串类型的混淆
- null与“null”的混淆

关于数据结构，后端也常有不按照约定返回时候。比如我们约定好返回的数据结构是这样的：

```json
{
    errorCode: 0,
    msg: '',
    data: {
        list: []
    }
}
```

结果，他给我返回的是：

```json
{
    errorCode: 0,
    data: {
        msg: '',
        data: []
    }
}
```

综上所述的种种情况，如果我们太过信任通信对方所返回的数据，而不进行相应的类型和字段防守的话，那么我们的JS代码极有可能在运行时报错TypeError。

## 前端错误捕获的方式

虽然，具备良好的防范意识和借助Typescript这个工具，能够把很多低级的错误扼杀在摇篮中。但是，最最坏的打算，终究是好的。出错不要紧，要紧的是要知道怎么处理。怎么处理呢？可以分为两个步骤：

- 第一步：捕获异常，获取错误信息
- 第二步：视实际需求而做对应的处理

### 第一步：捕获异常，获取错误信息

在页面中，捕获异常的技术手段无非有以下几种：

* try...catch
* window.onerror
* window.addEventListener('error',()=>{})
* element.onerror
* Promise Catch与window.addEventListener("unhandledrejection",()=> {})
* iframe与iframe.onload
* 其他

下面一一进行详细讲解。

#### try ... catch

(1) 动机

使用try...catch来捕获异常，归纳起来主要有两个动机：1）是真真正正地想对可能发生错误的代码进行异常捕获；2）想保证后面的代码继续运行

(2) [语法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch)

```js
try {
   try_statements
}
[catch (exception_var_1 if condition_1) { // non-standard
   catch_statements_1
}]
...
[catch (exception_var_2) {
   catch_statements_2
}]
[finally {
   finally_statements
}]
```

也就是说有以下几种组合方式：

* try...catch
* try...finally
* try...catch...finally

使用try...catch的时候，语法上有以下几个注意点：

- catch子句会收到一个错误对象。即使你不想使用这个错误对象，你也必须给它一个名字，否则浏览器会报一个“Uncaught SyntaxError”的错误。
- finally子句无论如何都会执行，并且优先级比return语句还要高。比如，下面这个函数会返回0，而不是2:

```js
function testFinally(){
    try{
        return 2;
    }catch(e){
        return 1;
    }finally{
        return 0;
    }
}
```

(3) 处理范围

只能捕获同步代码所产生的运行时错误，对于语法错误和异步代码所产生的错误是无能为力的。

#### window.onerror

(1) 动机

众所周知，很多做错误监控和上报的类库就是基于这个特性来实现的，我们期待它能处理那些try...catch不能处理的错误。

(2) 语法

```js
window.onerror = function(message, source, lineno, colno, error) { ... }
```

历史原因，不同浏览器或者同一个浏览器的不同版本中，window.onerror事件处理函数所接收到的参数也是不同的。这里只是讨论标准规范：

* message: 错误信息的字符串（string）
* source: 出错js文件的URL（string）
* lineno: 出错代码的行号（number）
* colno: 出错代码的列号（number）
* error: [错误对象](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error)（object）

就像上面说的，因为传入onerror事件处理函数的参数除了message外，在跨浏览器上具有兼容性问题，所以，如果考虑浏览器兼容性，建议只使用message参数即可。

这里值得提一点是，如果想要window.onerror捕获错误后，取消它的默认行为（这个默认行为就是在浏览器控制台打印出一个uncaught xxxxError ）,`事件处理器需要返回true`。这是反常识的。因为在UI事件系统中，取消默认行为一般是通过返回false来完成。

(3) 范围

3.1 window.onerror不能捕获语法错误

3.2  window.onerror不能捕获异步代码的错误

3.3  **但是有一点值得注意的是，window.onerror竟然能捕获同样是异步代码的setTimeout和setInterval里面的错误**：

```js
window.onerror = function(message, source, lineno, colno, error) {
    console.log('通过window.onerror捕获到的异常：', message);
    return true;
}

setTimeout(()=> {
    console.log(a);
},0)

setInterval(() => {
    console.log(b);
 }, 1000);

```

3.4  window.onerror 不能捕获静态资源的加载失败

3.5  window.oneror是能捕获**发生在事件处理函数里面的**包括语法错误在内的异常

```js
<script>
 window.onerror = function(message, source, lineno, colno, error) {
     console.log('通过window.onerror捕获到的异常：',message);
     return true;
 }
 </script>

<img id="img" src="./fake.png" onerror="console.log(abc)">
```

控制台会打印出：

> 通过window.onerror捕获到的异常： Uncaught ReferenceError: abc is not defined

看来，window.onerror并不是万能的。我们还指望它完完全全地去兜住try...catch的底，结果是它也有它捕获不了的错误。

#### window.addEventListener('error',()=>{})和 element.onerror

(1) 动机

因为window.onerror不能捕获网络资源加载失败的错误，我们希望这两个技术手段能满足我们的需求。

(2) 语法

就像[MDN](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror)所说的:

> When a resource (such as an <img> or<script>) fails to load, an error event using interface Event is fired at the element that initiated the load, and the onerror() handler on the element is invoked. These error events do not bubble up to window, but (at least in Firefox) can be handled with a window.addEventListener configured with useCapture set to True.

可以看出，这两者是在同一个事件传播路径上的：一个是target phase，一个capture phase。通过设置window.addEventListener()的第二个参数为true，是能够捕获网络资源加载失败的这种类型的错误的。

(3) 处理范围

我们看一个例子来看看，window.addEventListener('error')是否能够真的符合我们的期待：

```js
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>网络资源加载错误捕获</title>
    <script>
        window.addEventListener('error', (error) => {
            console.log('使用window.addEventListener能捕获到异常：', error);
        }, true)
    </script>
    <script src="./fake.js"></script>
    <link rel="stylesheet" href="./fake.css">
</head>
<body>
    <img id="img" src="./fake.png">
    <video id="video" src="fake.mp4"></video>
    <audio src="fake.mp3"></audio>
    <iframe id="iframe" src="./test4.html" frameborder="0" ></iframe>
</body>
</html>

```

下面我们来看看控制台打印结果：

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghoe9syui1j314c0jy7cb.jpg)

可以看到，除了iframe的所有资源的加载错误都被window.addEventListener('error')所捕获到了。

看来iframe要区别对待来行啊。

#### Promise Catch与window.addEventListener("unhandledrejection",()=> {})

(1) 动机

目前，异步代码中，除了promise代码错误捕获没有被讲到之外，setTimeout/setInterval是能被window.onerror所捕获的，async...await是能被try...catch所捕获的。那下面，我们来看看promise代码中的错误应该如何捕获。

(2) 语法

```js
const promiseObj = new Promise(executor);

promiseObj
.then(handleFulfilledA,handleRejectedA)
.then(handleFulfilledB,handleRejectedB)
.then(handleFulfilledC,handleRejectedC);

// 或者
promiseObj
.then(handleFulfilledA)
.then(handleFulfilledB)
.then(handleFulfilledC)
.catch(handleRejectedAny);
```

MDN更推荐后者的写法。这种写法的语法风格更接近try...catch，更易于接受。

(3) 处理范围

目前来看，没有任何的技术手段来捕获promise所产生的异常，promise自带的catch来得正是时候。

```js
const p = new Promise(res => {console.log(a)});
p.then(() => {}).catch(e => {
  console.log('Promise catch捕获的异常: ', e);
});
```

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghmk8dpj1lj30tw04cwfk.jpg)

是的，是没问题的。但是，有时候无论是自己还是同事都会忘记写catch分支，这个时候在全局增加一个对 unhandledrejection 事件的监听来进行兜底，无疑是明智之举。

```js
window.addEventListener("unhandledrejection", function(e){
    console.log('捕获到的promise异常:', e);
    e.preventDefault();
});
new Promise((res)=>{console.log(a)});
```

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghmkfdeo12j312m0j6afm.jpg)

这里值得一提的是，不同于DOM1用`return true`，DOM3级的事件机制是用`event.preventDefault()`来去取消事件的默认行为的。同样，在这里的默认行为是“在浏览器控制台打印出一个uncaught xxxxError”。

#### iframe与iframe.onload

上面在window.addEventListener('error')一小节中说到，除了iframe以外，其他类型的网络静态资源的加载错误都能够被它所捕获。那剩下个iframe，我们该怎么办呢?

需要采用比较hack的方法：

加载失败后，iframe所加载页面的文档title都会被设置为“Error”

最后，结合window.onerror所实现的方案是这样的：

```js
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>iframe加载错误</title>
</head>

<body>
    <iframe id="iframe" src="./test123.html" frameborder="0">
    </iframe>
    <script>
        window.onerror = function (message, source, lineno, colno, error) {
            console.log('通过window.onerror捕获到的异常：', message);
            return true;
        }

        var iframe = document.getElementById('iframe');
        iframe.onload = function (event) {
            var title = this.contentWindow.document.title.toLowerCase();
            // console.log('错误标题 ', title);
            if (title === 'error') {
                throw new Error('iframe加载失败');
            }
        }
    </script>
</body>

</html>
```

结果是控制台会打印出：

> 通过window.onerror捕获到的异常： Uncaught Error: iframe加载失败

#### 其他

##### 1. page crash

借助 **beforeunload**事件 和 **service worker**实现的两种监控网页崩溃的文案，具体可参考[如何监控网页崩溃?](https://juejin.im/entry/6844903705318260743?utm_source=gold_browser_extension)这篇文章。

##### 2. script error

通常为了捕获此类错误需要进行两步操作：

第一步：在script标签上增加`crossorigin`属性

第二步：在response headers中增加`Access-Control-Allow-Origin`

详见[处理error: Script error报错](http://confluence.haodf.net/pages/viewpage.action?pageId=20740455)

### 第二步：视实际需求而做对应的处理

上面的第一步其实就是讲如何针对各种错误场景去找到其合适的处理方案。处理的目的无非就是为了知道当前出错了或者出现了什么错误。

那么第二步就要讲如何处置这些错误信息。这无非也就包括以下两种方式：

* 把错误信息上报给前端监控系统
* 通过界面反馈，告知用户当前页面出错了

##### 把错误信息上报给前端监控系统

把错误信息上报给监控系统之前，往往需要对错误进行定级。如何定级，这是一个因人而异，又或者因业务而异的事情。一般情况下，我们基本上可以划分为两级：致命跟非致命的。

非致命错误，可以根据以下一个或者多个条件来确定：

* 不影响用户的主要任务
* 只影响页面的一部分
* 可以恢复
* 重复操作可以消除错误

致命错误，可以通过以下一个或多个条件来确定：

* 应用程序根本无法继续运行
* 错误明显影响到用户的主要操作
* 会导致其他连带的错误

一般而言，像图片加载失败啊，因为网络原因导致API请求超时等等，这些都属于非致命的错误。基本可以这么说，非致命的错误的并不是需要主要关注的问题。我们在监控系统排查和定位问题的时候，我们首要关注的肯定是致命问题。

一般而言，没有主动通过try...catch来捕获，而是最终被兜底的window.onerror来捕获的错误，很有可能会导致应用程序无法继续运行或者影响用户主要操作。这类型的错误得优先排查和修复。

在某种场景下，如果必须刷新页面才能让应用程序继续正常运行的话，那就必须通知用户，同时给用户提供要给点击即可刷新页面的按钮。

确定好错误的等级，再加上拿到的错误信息，我们就可以上报给监控系统了。

常用方案就是使用Image对象来发送请求。这么做，有以下几点好处：

* 浏览器兼容性好。所有浏览器都支持Image对象。
* 可以避免same-origin-policy的限制。实际开发中，通常都是一台服务器负责接收所有的服务器的错误上报。而这种情况下，就会存在跨域问题，单纯的XMLHttpRequest是解决不了这个问题。
* 在上报过程上，本身出现错误的概率比较低。如果你自己封装了一个ajax库或者使用了外部ajax库，如果这些库本身就有问题的话，你还指望它去上报的话，可想而知，结果是无法上报成功的

下面，我们简单实现一个用于上报错误的方法：

```js
function postError(type, msg) {
    const img = new Image();
    img.src = `log.php?type=${encodeURICoponent(type)}&msg=${encodeURICoponent(msg)}`
}
```

##### 通过界面反馈，告知用户当前页面出错了

当页面出现错误了，尤其是出现了那些估计会导致用户点击页面没反应的错误的时候，一定要给出一些友好的提示，尊重用户的知情权。

提示语首先得友好，用户能读懂。其次是相对准确。提示的表现形式一般有轻提示toast，重提示弹窗，局部页面替换和整个页面替换（类似于常见的404错误提示）等几种。具体采取哪种形式，可以根据具体的业务场景和错误等级来决定。

一般而言，错误上报和界面反馈最好是结合起来使用，双管齐下，取得的生产效果是最佳。

## 总结

以一张图来总结本文的核心内容：

![](https://tva1.sinaimg.cn/large/007S8ZIlgy1ghofjg6gybj32400nuwqt.jpg)

## 参考资料

1. JavaScript高级程序设计（第3版）第17章
2. DOMException [https://developers.google.cn/web/updates/2017/06/play-request-was-interrupted?hl=en#fix](https://developers.google.cn/web/updates/2017/06/play-request-was-interrupted?hl=en#fix)
3. DOMException [http://help.dottoro.com/ljushpcx.php;](http://help.dottoro.com/ljushpcx.php;)
4. try...catch [https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/try...catch](https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/try...catch)
5. windw.onerror [https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onerror)
